src/main/archetype/core/src/test/java/core/models/MyProductTeaserImplTest.java

```diff
@@ -0,0 +1,132 @@
+/*******************************************************************************
+ *
+ *    Copyright 2020 Adobe. All rights reserved.
+ *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License. You may obtain a copy
+ *    of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software distributed under
+ *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ *    OF ANY KIND, either express or implied. See the License for the specific language
+ *    governing permissions and limitations under the License.
+ *
+ ******************************************************************************/
+package ${package}.core.models;
+
+import com.adobe.cq.commerce.core.components.models.productteaser.ProductTeaser;
+import com.adobe.cq.commerce.core.components.models.retriever.AbstractProductRetriever;
+import com.adobe.cq.commerce.magento.graphql.ProductInterface;
+import junit.framework.Assert;
+import org.apache.sling.api.resource.ValueMap;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+class MyProductTeaserImplTest {
```

Comment:
I'm fine with the mocking of the delegate methods. For the `MyProductTeaser` specific implementation however, I would like to see tests using `AemContext` like in the components project, to make sure the component actually works.

https://github.com/adobe/aem-core-cif-components/blob/master/bundles/core/src/test/java/com/adobe/cq/commerce/core/components/internal/models/v1/productteaser/ProductTeaserImplTest.java#L52

Reply:
That would be more a functional test rather than unit test. However I tried using `AemContext`, load specific content and `adaptTo(MyProductTeaser.class) `but that would add dependency on the `ProductTeaser` internal implementation. For instance current (and not released) `ProductTeaser` model uses `UrlProvider` which needs to be mocked for the injections to work. This implies the custom code needs to be aware of the core components internals in order to write successful unit tests using `adaptTo`

Reply:
You can mock the injection of https://github.com/adobe/aem-cif-project-archetype/blob/master/src/main/archetype/core/src/main/java/core/models/MyProductTeaserImpl.java#L41-L43 via https://github.com/adobe/aem-core-cif-components/blob/master/bundles/core/src/test/java/com/adobe/cq/commerce/core/components/internal/models/v1/product/ProductImplTest.java#L117-L121 and then you can mock the delegated methods like before.

Reply:
You can also look at the tests in https://github.com/adobe/aem-core-cif-components/blob/master/examples/bundle/src/test/java/com/adobe/cq/commerce/core/examples/servlets/GraphqlServletTest.java where it adapts to the `interface` classes, not the `impl`. I agree with @hbacila that our tests are more than unit tests, but I think it has a real benefit because we do not yet have integration tests.

Reply:
I forgot to mention that I got issues getting this to work, see https://github.com/adobe/aem-core-cif-components/blob/master/examples/bundle/pom.xml#L401-L407 - we need at least this version so that the injection properly works with `@Self` and `@Via`.

Reply:
also the archetype uses the latest JUnit5. Core components still use Vintage JUnit.
Will look into it.

Reply:
@herzog31 @cjelger I refactored the test. Please have a look. Thanks